go/interpreter/lfvm/memory_test.go

```diff
@@ -47,3 +49,134 @@ func TestExpansionCosts(t *testing.T) {
+		},
+		"memory-bigger-than-size-changes-nothing": {
+			size:        32,
+			initialMem:  []byte{63: 0x0},
+			expectedMem: []byte{63: 0x0},
+		},
+		"size-zero-changes-nothing": {
+			size:        0,
+			initialMem:  []byte{31: 0x0},
+			expectedMem: []byte{31: 0x0},
+		},
+	}
+
+	for name, test := range test {
+		t.Run(name, func(t *testing.T) {
+			m := NewMemory()
+			fee := m.ExpansionCosts(test.size)
+			m.store = test.initialMem
+			m.expandMemoryWithoutCharging(test.size, fee)
+			if !bytes.Equal(m.store, test.expectedMem) {
+				t.Errorf("unexpected memory value, want: %x, got: %x", test.expectedMem, m.store)
+			}
+			if m.total_memory_cost != fee {
+				t.Errorf("unexpected total memory cost, want: %d, got: %d", fee, m.total_memory_cost)
+			}
+		})
+	}
+
+}
+
+func TestMemory_expandMemory_ErrorCases(t *testing.T) {
+
+	tests := map[string]struct {
+		size     uint64
+		offset   uint64
+		gas      tosca.Gas
+		expected error
+	}{
+		"not enough gas": {
+			size:     32,
+			offset:   0,
+			gas:      0,
+			expected: errOutOfGas,
+		},
+		"offset overflow": {
+			size:     1,
+			offset:   math.MaxUint64,
+			gas:      100,
+			expected: errGasUintOverflow,
+		},
+		"size overflow": {
+			size:     math.MaxUint64,
+			offset:   1,
+			gas:      100,
+			expected: errGasUintOverflow,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			ctxt := getEmptyContext()
+			m := NewMemory()
+			ctxt.gas = test.gas
+
+			err := m.expandMemory(test.offset, test.size, &ctxt)
+			if !errors.Is(err, test.expected) {
+				t.Errorf("unexpected error, want: %v, got: %v", test.expected, err)
+			}
+			if m.Len() != 0 {
+				t.Errorf("should have remained empty, got length: %d", m.Len())
+			}
+		})
+	}
+}
+
+func TestMemory_expandMemory(t *testing.T) {
+
+	tests := map[string]struct {
+		size          uint64
+		offset        uint64
+		initialMemory []byte
+	}{
+		"empty-memory-with-zero-offset-and-size-does-not-expand": {},
+		"size-zero-with-offset-does-not-expand": {
+			size:   0,
+			offset: 32,
+		},
+		"expand-memory": {
+			size:   32,
+			offset: 0,
+		},
+		"memory-bigger-than-size+offset-does-not-expand": {
+			size:          32,
+			offset:        64,
+			initialMemory: []byte{127: 0x0},
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			ctxt := getEmptyContext()
+			m := NewMemory()
+			ctxt.gas = 100
+
+			err := m.expandMemory(test.offset, test.size, &ctxt)
+			memSize := m.Len()
+			if err != nil {
+				t.Fatalf("unexpected error, want: %v, got: %v", nil, err)
+			}
+			// we split memory size check in two cases because when size is zero,
+			// we don't expect the memory to be expanded, regardless of the offset
+			if test.size == 0 && m.Len() != uint64(len(test.initialMemory)) {
+				t.Errorf("memory size should not have changed, want: %d, got: %d", len(test.initialMemory), memSize)
+			} else if want := test.size + test.offset; test.size != 0 && memSize < want {
```

Comment:
Is it really required for the memory to be greater and not just greater equal?

Reply:
yes. because memory expands to exactly `offset+size`, or if existing memory was greater than that addition, then it should remain bigger.